L2 Health Checks

amarjandu opened the pull request:
<!-- Please make sure to provide a meaningful title for your PR. Do not keep the default title. -->
<!-- Use the following GitHub keyword if your PR completely resolves an issue: "Fixes #123" -->

Implements internal health checks for the DSS
<!-- Describe how you tested this PR, and how you will test the feature after it is merged. Uncomment below:
### Test plan -->

<!-- Describe any special instructions to the operator who will deploy your code to the integration, 
     staging and production deployments. Uncomment below:
### Deployment instructions & migrations -->

<!-- Add notes to highlight the feature when it's released/demoed. Uncomment the headings below:
### Release notes -->

<!-- Do you want your PR to be merged by the reviewer using squash or rebase merging? If so, mention it here. -->

xbrianh commented on daemons/dss-canary/app.py:
```diff
@@ -0,0 +1,43 @@
+canaryKey = 'canary/health'
+
+
+@app.scheduled_function('rate({} minute)'.format(delete_timer))
```
Instead of creating a new daemon, consider using the GCP Cloud Schedule, which can inject messages into the dss-gs-event-relay pub/sub topic. This will not require the creation of a canary file: https://cloud.google.com/scheduler/docs/

amarjandu replied:
@xbrianh i ended up removing the Canary and just keeping the check to specifically the Event Relay.

Bento007 commented on health:
```diff
@@ -0,0 +1 @@
+{"Healthy": false, "canary": {"ElapsedTimeMinutes": 1405.9830637166667, "LastModified": "Thu, 31 Jan 2019 21:56:15 GMT"}}
```
What is this used for?

amarjandu replied:
removed,

xbrianh commented on dss/api/health.py:
```diff
@@ -0,0 +1,78 @@
+def _get_dynamodb_status():
+    db_status = True
+    ddb_tables = list(filter(lambda x: x.startswith("dss") and x.endswith(os.getenv("DSS_DEPLOYMENT_STAGE")),
+                             dynamodb.list_tables()["TableNames"]))
```
The criteria used to find dynamo-db tables is not exclusive to the data-store. For now, perhaps we should use an explicit list.

amarjandu replied:
changed to explicit list

xbrianh commented on dss/api/health.py:
```diff
@@ -0,0 +1,78 @@
+        if er_res['status'] == 'ACTIVE':
+            er_status = True
+            er_data = er_res
+    except Exception:
```
Catch-all try-except blocks may hide exceptions we wish to see.

Is there a narrower exception we can use?

amarjandu commented on chalice/app.py:
```diff
@@ -236,10 +237,11 @@ def version():
-                                headers={"Content-Type": "text/html"},
-                                body="OK")
+                                headers={"Content-Type": "application/json"},
+                                body=health_status)
```
placed the logging into `dss/api/health.py/l2_health_check()`

amarjandu commented:
@xbrianh  i changed the response from a failure to what we spoke about earlier in the week, to just have a `{"Healthy" :False}`, and moving the actual failure status into the log. 
Also added changes to have more specific exception failures within the tests.
